### DISCUSSION TITLE
I dont see apt plugin support processor in classes from other maven module ( multimodule proj)

### DISCUSSION THREAD

**Discussion by mpet** [eclipse-m2e/m2e-core#655] [Q&A] [answered]
Hi,

I managed to find the class where this takes place:

AbstractAptConfiguratorDelegate. configureProject(IProgressMonitor monitor)

I can see that the class AnnotationProcessorConfiguration contains the project with my annotation processor.

```
[C:\Users\eraonel\git\tutorials\annotations\annotation-processing\target\classes,
C:\repo\eraonel.m2\repository\org\slf4j\slf4j-api\1.7.32\slf4j-api-1.7.32.jar,
C:\repo\eraonel.m2\repository\ch\qos\logback\logback-classic\1.2.6\logback-classic-1.2.6.jar,
C:\repo\eraonel.m2\repository\ch\qos\logback\logback-core\1.2.6\logback-core-1.2.6.jar,
C:\repo\eraonel.m2\repository\org\slf4j\jcl-over-slf4j\1.7.32\jcl-over-slf4j-1.7.32.jar]
```
So far so good.

I want this to be searched for the processor but I don't see that happen since it is not a jar file.

```
// Inspect the dependencies to see if any contain APT processors
    boolean isAnnotationProcessingEnabled = configuration.isAnnotationProcessingEnabled()
        && containsAptProcessors(resolvedJarArtifacts);
```
This will check that I have a processor but will not find my classes since resolvedJarArtifacts only contains 

```
//The plugin dependencies are added first to the classpath
    LinkedHashSet<File> resolvedJarArtifacts = new LinkedHashSet<>(getJars(configuration.getDependencies()));
```

So from here resolvedJarArtifacts will only contain jar dependencies in the list showed in the beginning.

1. Am I correct that we lack support for finding annatations processors as a maven module ?
2. Is there a way for a plugin to find annotation processors in target/clasesses?

I would really appreciate your support and insights on this matter. Looking forward to discuss it.

//mike

**Comment by laeubi** [comment]
If it could be loaded from a jar, then it could also be loaded from `target/clasesses` I think this is just a simplification, as you need to be careful when using this:

1. you should not keep open files on the classfiles as the might need to be recreated
2. you need to think what should happen if a classfile changes? Do nothing and wait for the next build? Scedule a new generation of the APT processor? ...

**Reply by mpet** [reply]
Thanks  @laeubi for taking your time about this!  So it basically complicates the situation to have the annotation-processor as a maven module in the same project?
But ok it is possible to use taget/classes but it takes more effort to add support since it is depends on a source file.

So if the annotation processor implementation ( class extending AbstractProcessor ) is changed should it not behave the same as all other source code? I  mean a new build should be done of annotation processor and classes in annotation user module needs to be regenerated. Do you agree?

**Comment by mpet** [comment]
When I do Maven --> Update I see that m2e-apt is triggered. A Project -- > Clean from within Eclipse does not trigger m2e-apt. I

For Maven --> Update  I can see the following options:

![image](https://user-images.githubusercontent.com/1445826/164894282-8c2f0512-a124-4902-8648-eaa57b8bb71b.png)

Two options are:
1. Refresh workspace resources from local file system
2. Clean projects 

So if the source of annotation processor classes changed these selections will mean a new version in target/classes. Right?
In that case it should be safe to add the target/classes and search for annotations processor as well. What do you think?

**Reply by laeubi** [reply]
I think your best chance is to try that out and get some experience with that :-)

**Reply by mpet** [reply]
I run Maven --> Update 

I have tried and I can see that when I change and make my annotation processor module a local jar and add it as a [<annotationProcessorPaths>](https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#annotationProcessorPaths)
then configurations is picked up and I can see the annotation-processor jar in my .factorypath

When eclipse compilation is done I cannot see see that the processor is executed at all. This possible to see when I run debug for 
MavenCompilerJdtAptDelegate 
After the configuration is set using

org.eclipse.jdt.apt.core.util.AptConfig.setFactoryPath(javaProject, factoryPath);

Then after this I can see the "regular" builder executing
```
MavenBuilder$2(MavenBuilder$BuildMethod<T>).execute(int, Map<String,String>, IProgressMonitor) line: 73	
MavenBuilder.clean(IProgressMonitor) line: 210	
BuildManager$2.run() line: 858	
SafeRunner.run(ISafeRunnable) line: 45	
BuildManager.basicBuild(int, IncrementalProjectBuilder, Map<String,String>, MultiStatus, IProgressMonitor) line: 232	
BuildManager.basicBuild(IBuildConfiguration, int, IBuildContext, ICommand[], MultiStatus, IProgressMonitor) line: 281	
BuildManager$1.run() line: 334	
SafeRunner.run(ISafeRunnable) line: 45	
BuildManager.basicBuild(IBuildConfiguration, int, IBuildContext, MultiStatus, IProgressMonitor) line: 337	
BuildManager.basicBuild(IBuildConfiguration, int, IBuildContext, IProgressMonitor) line: 296	
BuildManager.build(IBuildConfiguration, int, String, Map<String,String>, IProgressMonitor) line: 488	
Project$1.run(IProgressMonitor) line: 565	
Workspace.run(ICoreRunnable, ISchedulingRule, int, IProgressMonitor) line: 2313	
Project.internalBuild(IBuildConfiguration, int, String, Map<String,String>, IProgressMonitor) line: 603	
Project.build(int, IProgressMonitor) line: 116	
ProjectConfigurationManager.updateProjectConfiguration0(Collection<IFile>, boolean, boolean, boolean, IProgressMonitor) line: 474	
ProjectConfigurationManager.lambda$1(MavenUpdateRequest, boolean, boolean, boolean, IMavenExecutionContext, IProgressMonitor) line: 362	
1734626671.call(IMavenExecutionContext, IProgressMonitor) line: not available	
MavenExecutionContext.executeBare(MavenProject, ICallable<V>, IProgressMonitor) line: 182	
MavenExecutionContext.execute(MavenProject, ICallable<V>, IProgressMonitor) line: 156	
MavenExecutionContext.execute(ICallable<V>, IProgressMonitor) line: 103	
MavenImpl.execute(boolean, boolean, ICallable<V>, IProgressMonitor) line: 1411	
ProjectConfigurationManager.updateProjectConfiguration(MavenUpdateRequest, boolean, boolean, boolean, IProgressMonitor) line: 361	
UpdateMavenProjectJob.runInWorkspace(IProgressMonitor) line: 80	
UpdateMavenProjectJob(InternalWorkspaceJob).run(IProgressMonitor) line: 42	
```
so no classes are generated.

It is basically dead after doing update. 

@fbricon I can see that you did a major update to this. Is is possible that you can explain how the properties is in
pom.xml for annotationProcessorPaths go together with the jdt-apt build in Eclipse?


Judging from all the questions  I found I guess I am not the only one that is confused over this 😄

**Comment by fbricon** [comment, marked as answer]
Long story short, JDT APT doesn't support Annotation Processors from workspace projects. They must come from jar files.
It's a design decision as explained in https://bugs.eclipse.org/bugs/show_bug.cgi?id=259230#c1. 

Your best bet is to make changes to your annotation processor in another Eclipse instance, and have the processor project closed in the workspace you want to consume it, so that it's referenced as a jar file.

**Reply by mpet** [reply]
@fbricon  Thanks for replying!!

I have have created a jar file of my annotation processor and add it as a jar dependency. So the m2e-apt plugin picks up the processor. I checked AnnotationServiceLocator.getAptServiceEntries() 
![image](https://user-images.githubusercontent.com/1445826/165104738-d53d7221-e207-4b97-a621-e9e2b3fd1b7f.png)

I also checked the .factorypath and it contains the jar with annotation processor.

Last thing I see from the m2e-apt plugin in AbstractAptConfiguratorDelegate is this line called:
 
// Apply that IFactoryPath to the project
    AptConfig.setFactoryPath(javaProject, factoryPath);

Then suddenly I can see that the UpdateMavenProjectJob calls:
MavenBuilder.execute and it does not generate any classes.

I was expecting that any of the jdt-apt classes was called for build but it don't see that happen.

So what do you expect to happen after a     AptConfig.setFactoryPath(javaProject, factoryPath) is called ?

**Reply by fbricon** [reply]
When jdt_apt is selected for performing AP, then m2e-apt no longer does anything other than configuring APT. The actual processing is 100% done by JDT APT. If things are not generated, it's possible something is wrong in the processor itself, but then you'd have to debug JDT APT yourself to figure out why.

https://github.com/eclipse/eclipse.jdt.ls/tree/master/org.eclipse.jdt.ls.tests/projects/maven/autovalued is an example of working AP

**Reply by mpet** [reply]
@fbricon when I `run mvn clean compile` the files are generated

What settings should I have for m2e-apt to make your example work since this seems to be the main issue.

I set the following for the test project.

![image](https://user-images.githubusercontent.com/1445826/165117842-43f8bf3f-5ea5-4bc5-85fd-60b312fe6c10.png)

And then I run Maven--> Update on the complete test project.
No generated files

**Reply by fbricon** [reply]
That's the proper setting. But no need to import all jdt.ls projects into Eclipse, this will not configure the test project. Remove them altogether and simply import org.eclipse.jdt.ls.tests/projects/maven/autovalued as a Maven project.

**Reply by mpet** [reply]
@fbricon  I ended up doing the following.

In my annotation-processor module I build a jar. Then close the module. And then I have a dependency looking like this in my processor-user module.
```
<dependencies>
        <dependency>
            <groupId>com.baeldung</groupId>
            <artifactId>annotation-processing</artifactId>
            <version>1.0.0-SNAPSHOT</version>
            <optional>true</optional>
        </dependency>
    </dependencies>
```
I also found out that this did not work in Eclipse 4.21 so I upgraded to Eclipse 4.23.
And it works like a charm. 

Many thanks for your support!
